core/src/main/java/org/elasticsearch/search/rescore/RescoreBuilder.java

```diff
@@ -78,23 +139,41 @@ public XContentBuilder toXContent(XContentBuilder builder, Params params) throws
 
         protected abstract XContentBuilder innerToXContent(XContentBuilder builder, Params params) throws IOException;
 
+        @Override
+        public abstract int hashCode();
+
+        @Override
+        public abstract boolean equals(Object obj);
     }
 
     public static class QueryRescorer extends Rescorer {
+
         private static final String NAME = "query";
-        private QueryBuilder queryBuilder;
-        private Float rescoreQueryWeight;
-        private Float queryWeight;
-        private String scoreMode;
+        public static final QueryRescorer PROTOTYPE = new QueryRescorer(new MatchAllQueryBuilder());
+        public static final float DEFAULT_RESCORE_QUERYWEIGHT = 1.0f;
+        public static final float DEFAULT_QUERYWEIGHT = 1.0f;
+        public static final QueryRescoreMode DEFAULT_SCORE_MODE = QueryRescoreMode.Total;
+        private QueryBuilder<?> queryBuilder;
```

Comment:
I think we should maybe try to solve that in ObjectParser as it would be better to have all compulsory things set in the constructor so that we don't end up with validation problems or esoteric bugs because it hasn't been set later. Maybe @s1monw already has a solution for this?